crates/typst-kit/src/package.rs

```diff
@@ -0,0 +1,168 @@
+//! Download and unpack packages and package indices.
+
+use std::fs;
+use std::path::{Path, PathBuf};
+
+use ecow::eco_format;
+use once_cell::sync::OnceCell;
+use typst::diag::{bail, PackageError, PackageResult, StrResult};
+use typst::syntax::package::{
+    PackageInfo, PackageSpec, PackageVersion, VersionlessPackageSpec,
+};
+
+use crate::download::{Downloader, Progress};
+
+/// The default Typst registry.
+pub const DEFAULT_REGISTRY: &str = "https://packages.typst.org";
+
+/// The default packages sub directory within the package and package cache paths.
+pub const DEFAULT_PACKAGES_SUBDIR: &str = "typst/packages";
+
+/// Holds information about where packages should be stored and downloads them
+/// on demand, if possible.
+#[derive(Debug)]
+pub struct PackageStorage {
+    /// The path at which non-local packages should be stored when downloaded.
+    pub package_cache_path: Option<PathBuf>,
+    /// The path at which local packages are stored.
+    pub package_path: Option<PathBuf>,
```

Comment:
perhaps we should keep these private for flexibility? or are they used externally?

Reply:
Well, they are used in the World impl in `typst-cli`, I figured, since we pass them in ourselves and don't have any restrictions regarding access (the consumer can read and write them), we can keep them public.

Reply:
What I'm a little cautious about is people changing them half-way through. Right now it's maybe ok, but we might not want to give that guarantee. 

That aside, I can't actually find where in the CLI World impl they are used.

Reply:
It seems they weren't, I went with getters only.